rats-tls/src/lib.rs

```diff
@@ -0,0 +1,239 @@
+
+unsafe impl Send for RatsTlsRef {}
+unsafe impl Sync for RatsTlsRef {}
+unsafe impl Send for RatsTls {}
+unsafe impl Sync for RatsTls {}
+
+unsafe impl ForeignType for RatsTls {
+    type CType = rats_tls_handle;
+    type Ref = RatsTlsRef;
+
+    unsafe fn from_ptr(ptr: *mut rats_tls_handle) -> RatsTls {
+        RatsTls(NonNull::new_unchecked(ptr))
+    }
+
+    fn as_ptr(&self) -> *mut rats_tls_handle {
+        self.0.as_ptr()
+    }
+
+    fn into_ptr(self) -> *mut rats_tls_handle {
+        let inner = self.as_ptr();
+        ::core::mem::forget(self);
+        inner
+    }
+}
+
+impl Drop for RatsTls {
+    fn drop(&mut self) {
+        unsafe {
+            rats_tls_cleanup(self.as_ptr());
+        }
+    }
+}
+
+impl Deref for RatsTls {
+    type Target = RatsTlsRef;
+
+    fn deref(&self) -> &RatsTlsRef {
+        unsafe { RatsTlsRef::from_ptr(self.as_ptr()) }
+    }
+}
+
+impl DerefMut for RatsTls {
+    fn deref_mut(&mut self) -> &mut RatsTlsRef {
+        unsafe { RatsTlsRef::from_ptr_mut(self.as_ptr()) }
+    }
+}
+
+impl RatsTls {
+    pub fn new(
+        server: bool,
+        enclave_id: u64,
+        tls_type: Option<&str>,
+        crypto: Option<&str>,
+        attester: Option<&str>,
+        verifier: Option<&str>,
+        mutual: bool,
+    ) -> Result<RatsTls, rats_tls_err_t> {
+        let mut conf: rats_tls_conf_t = Default::default();
+        conf.api_version = RATS_TLS_API_VERSION_DEFAULT;
+        conf.log_level = RATS_TLS_LOG_LEVEL_DEBUG;
+        if let Some(tls_type) = tls_type {
+            conf.tls_type[..tls_type.len()].copy_from_slice(tls_type.as_bytes());
+        }
+        if let Some(crypto) = crypto {
+            conf.crypto_type[..crypto.len()].copy_from_slice(crypto.as_bytes());
+        }
+        if let Some(attester) = attester {
+            conf.attester_type[..attester.len()].copy_from_slice(attester.as_bytes());
+        }
+        if let Some(verifier) = verifier {
+            conf.verifier_type[..verifier.len()].copy_from_slice(verifier.as_bytes());
+        }
+        conf.cert_algo = RATS_TLS_CERT_ALGO_DEFAULT;
+        conf.enclave_id = enclave_id;
+        if mutual {
+            conf.flags |= RATS_TLS_CONF_FLAGS_MUTUAL;
+        }
+        if server {
+            conf.flags |= RATS_TLS_CONF_FLAGS_SERVER;
+        }
+
+        let mut handle: rats_tls_handle = unsafe { std::mem::zeroed() };
+        let mut tls: *mut rats_tls_handle = &mut handle;
+        let err = unsafe { rats_tls_init(&conf, &mut tls) };
+        if err != RATS_TLS_ERR_NONE {
+            // error!("rats_tls_init() failed");
+            return Err(err);
+        }
+
+        let err = unsafe { rats_tls_set_verification_callback(&mut tls, None) };
+        if err == RATS_TLS_ERR_NONE {
+            Ok(unsafe { RatsTls::from_ptr(tls) })
+        } else {
+            Err(err)
+        }
+    }
+
+    pub async fn negotiate_async(self, stream: TcpStream) -> std::io::Result<DuplexStream> {
```

Comment:
I like the async wrapper here, which will be of benefit to the performance, without blocking